### ISSUE TITLE
First character dropped from input

### ISSUE THREAD

**Issue by jimmyfrasche** [tdewolff/parse#8]
Given the program

```
package main

import (
	"log"
	"os"

	"github.com/tdewolff/parse/css"
)

func main() {
	ss, err := css.Parse(os.Stdin)
	if err != nil {
		log.Fatalln(err)
	}
	ss.Serialize(os.Stdout)
}
```

and the input

```
div{}
div{}
```

I would expect 

```
div{}div{}
```

but running

```
go run test.go <test.css
```

I get 

```
iv{}iv{}
```

Inspecting the values it seems that the first byte of the input is always read as ``\n`` (given a multibyte  utf-8 rune it will truncate the first byte but keep the rest)

edit: I fail at markdown

**Comment by tdewolff** [comment]
Strange, I cannot reproduce your result, for I obtain `div{}div{}`.

Are you sure you updated all the packages? Last night I did a major refactoring of mainly the parse/css package, but parse and minify/css have been touched as well.

**Comment by jimmyfrasche** [comment]
I didn't run into this until I updated. I don't see anything obvious in the updates.

I just ran another test where I typed ``div{}div{}^D`` and got ``di{}div{}`` back instead of ``iv{}div{}``.  I can reproduce in this different shells and terminals but that is plainly insane. I'm not sure what's going on: I think my computer has contracted a ghost.

```
$ cd github.com/tdewolff
$ go get -u ./...
$ go clean ./...
$ go install ./...
$ go test ./...
ok  	github.com/tdewolff/minify	0.022s
ok  	github.com/tdewolff/minify/css	0.010s
ok  	github.com/tdewolff/minify/html	0.012s
ok  	github.com/tdewolff/minify/trim	0.004s
ok  	github.com/tdewolff/parse	0.008s
ok  	github.com/tdewolff/parse/css	0.005s
ok  	github.com/tdewolff/parse/html	0.009s
$ go version
go version go1.4.2 linux/amd64
$ uname -roms
Linux 3.16.0-4-amd64 x86_64 GNU/Linux
$ printenv LANG
en_US.UTF-8
```

**Comment by tdewolff** [comment]
No, you were right. There is a memory bug somewhere and I suspect the internal `p.buf` is the culprit!

**Comment by jimmyfrasche** [comment]
Wild that it would deterministically manifest itself differently based solely on whether stdin was redirected. I saw that and knew I'd surely gone insane.

**Comment by tdewolff** [comment]
It behaves differently because if the source was not redirected (ie. it's in memory) the tokenizer (or rather `ShiftBuffer`) uses this in-memory buffer instead of reading it like it was a stream.

**Comment by tdewolff** [comment]
Fixed in 43ef46d4f81c2ce4d9688757da1e45cc699153cb, but it means giving up an optimization for safety. Will need to rethink buffer sharing...
